File: LeetCode/Graph/BFS/Golang/Leetcode886-possible-bipartition.go
package sbgo

func possibleBipartition(n int, dislikes [][]int) bool {
	g := make([][]int, n+1)
	for _, d := range dislikes {
		g[d[0]] = append(g[d[0]], d[1])
		g[d[1]] = append(g[d[1]], d[0])
	}
	color := make([]int, n+1)
	var dfs func(i, j int) bool
	dfs = func(u, c int) bool {
		color[u] = c
		for _, v := range g[u] {
			if color[v] != 0 {
				if color[v] == color[u] {
					return false
				}
			} else if !dfs(v, 3-c) {
				return false
			}
		}
		return true
	}
	for i := 1; i <= n; i++ {
		if color[i] == 0 && !dfs(i, 1) {
			return false
		}
	}
	return true
}

File: LeetCode/Graph/BFS/Python/Leetcode886-possible-bipartition.py
class Solution:
    def possibleBipartition(self, n: int, dislikes) -> bool:
        g = [[] for _ in range(n + 1)]
        for d in dislikes:
            g[d[0]].append(d[1])
            g[d[1]].append(d[0])

        color = [0] * (n + 1)
        def dfs(u,c):
            color[u] = c
            for v in g[u]:
                if color[v] != 0:
                    if color[v] == color[u]:
                        return False
                elif not dfs(v, 3 - c):
                    return False
            return True 

        for i in range(1, n + 1):
            if color[i] == 0 and not dfs(i,1):
                return False
        return True